### PR TITLE
anki: 2.1.15 -> 2.1.40 (fixes #112849) [DRAFT for others to pick up]

### LINKED ISSUE #112849: Anki 2.1.40
###### Checklist

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

###### Project name
`nix search` name: anki

current version: 2.1.15
desired version: 2.1.40

###### Notify maintainers

maintainers: @oxij @Profpatsch

###### Note for maintainers

Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by iblech**
###### Motivation for this change

#112849

###### Things done

I tackled some of the required changes, mostly owing to upstream's directory reorganization. However, I then noticed that upstream has switched to a new build system (bazel). Right now I unfortunately don't have the time to deal with the changes required by this switch.

**Note** that at the moment, this pull request is **not** fit for merging! The main executable does not start.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 112921` at 9501094a run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/5071da262ade041b963e89181a1eb939">ankisyncd</a></li></li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/4b25d70bc851c4736cf01bc0a67c2ac5">anki</a></li></li>
    <li><a href="https://gist.github.com/2009c9890e73d06781626f81e70254a4">mnemosyne</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `fetchpatch`.
Near pkgs/games/anki/default.nix:7:3:
```
  |
7 | , fetchpatch
  |   ^
```
</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 112921` at 9501094a run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/03bc068dbbb5900bd601ea74481154f7">ankisyncd</a></li></li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/61fa8ef1f21bc8983346dc7f18146056">anki</a></li></li>
    <li><a href="https://gist.github.com/c748ec0425c5536ed51e02e2abceaf7b">mnemosyne</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `fetchpatch`.
Near pkgs/games/anki/default.nix:7:3:
```
  |
7 | , fetchpatch
  |   ^
```
</li>
  </ul>
</details>

File: pkgs/games/anki/default.nix [hunk 1]
```diff
@@ -111,37 +113,23 @@ buildPythonApplication rec {
   patches = [
     # Disable updated version check.
     ./no-version-check.patch
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/games/anki/default.nix:114]: Please remove fetchpatch from inputs.

Comment (iblech) [reply pkgs/games/anki/default.nix:114]: Ah yes, thanks, did so.

**Comment by nessdoor** [comment]
I'm picking this up. Trying to fight Bazel and its obscurities, atm.

Just to let other people know.

**Comment by yuuyins** [comment]
Any luck? I may try it as well...

**Comment by ArdaXi** [comment]
For what it's worth, a while back I spent some time getting 2.1.44 working by taking out Bazel altogether. It was incredibly frustrating, and at some point I chose to build some stuff manually and pull other stuff out of the binary build because I just couldn't any more.

I am in no way proud of this, but if anyone wants to look at it, [it's in my overlay](https://github.com/ArdaXi/nixos/tree/flake/pkgs/anki). This is not a place of honor, etc etc.

**Comment by nessdoor** [comment]
@yuuyins unfortunately I am stuck at a point where Bazel complains about unspecified targets and I really have no idea how to debug the thing without taking a deeper dive into the query language or whatever else.

A quick rundown of my work so far:
- attempted to build on NixOS outside of Nix evaluation, by direct invocation of Bazel
- plugged-in Tweag-made rules for getting Nix-supplied binaries
- replaced CC, Rust, Python, Node and Yarn toolchains with binaries provided through Nix

Now, at this point all issued related to dynamic linking and whatnot seem to be fixed, and the build progresses well for a while until:
```
error: Error loading target specification: Could not find specification for target "". Use `--print target-list` for a list of built-in targets

ERROR: /home/tlopez/.cache/bazel/_bazel_tlopez/29596eeeaeb1c96a789bf8decf5825e8/external/io_bazel_rules_rust/cargo/cargo_build_script_runner/BUILD:3:13: output 'external/io_bazel_rules_rust/cargo/cargo_build_script_runner/libcargo_build_s\
cript_output_parser--359502323.rlib' was not created
ERROR: /home/tlopez/.cache/bazel/_bazel_tlopez/29596eeeaeb1c96a789bf8decf5825e8/external/io_bazel_rules_rust/cargo/cargo_build_script_runner/BUILD:3:13: Compiling Rust rlib cargo_build_script_output_parser (1 files) failed: not all output\
s were created or valid
```
and I swear, I have no clue what this means. I don't know if I got lost in something silly, or I have some methodological issue that prevents me from discovering the cause, or I have to blame the fact that I usually work on these things late at night, but I cannot get past this.

I am sorry to say that the last time I worked on this was on August 1st, then other things got in the way and I couldn't dedicate much time to solve the issue. Maybe I should really get back to it.

If you want to give it a try, I can send you the Bazel recipes that I came up with, so that you don't have to go through hell from the beginning. Mind you, I tried to build inside  a simple FHS environment as suggested by the NixOS wiki but to no avail. I don't know, maybe it's still  a viable approach, but I wanted to get a cleaner solution anyway.

One thing to note and that made me go a little crazy: the folks at Ankitects, for who knows what reason, use custom Bazel rules and a lot of precompiled binaries pulled directly from their own repos. And they are _not always_ a mirror of upstream rules. For example, their Rust ruleset comes with the persistent worker functionality that is their own addition (although their default build rule explicitly disables it). They make extensive use of replacement patterns to substitute upstream rules with their custom Bazel rule repos without changing the name, so it's not super obvious where the custom rules are actually used in stead of the upstream ones.

I hope someone well-versed in Bazel could help. I'll try to resume the efforts, sorry for the hiatus.

@ArdaXi tomorrow I will give a look at your approach. Maybe purging Bazel out of the way is a viable and most sane solution.

**Comment by ArdaXi** [comment]
@nessdoor Your progress looks similar to mine. I also tried the FHS approach, and I'm honestly kinda impressed you made it so far as to get Yarn working. I guess I've just not had the cognitive bandwidth to wrap my head around Bazel enough to use the Tweag rules.

I haven't worked on this since… June, so my memory's a bit fuzzy, but as I recall I only ended up copying the `fluent.proto` into the repository because it seemed the way it's handled would change in `2.1.45` so figuring out how that's done by hand didn't seem worth the effort.

The main thing that should really be improved (other than my terrible uncommented Bash) is the line

```
cp -rv ${anki-bin}/bin/aqt_data/web $pp/aqt/data/web
```

I remember spending a bunch of effort trying to figure out how those files are built, but I was also just trying to solve a problem, and this solved my problem. It's quite clearly unacceptable as a real solution, though.

Oh, since I don't think I actually have a license on that repo yet, let me just state for the record that to the extent possible under law, I've licensed the `pkgs/anki` subdirectory CC0 and waived all copyright and related or neighboring rights to the `pkgs/anki` subdirectory. Just in case anyone's concerned, I'll fix that properly later.

**Comment by nessdoor** [comment]
Ok, sorry for the delay, other things got in the way.

@ArdaXi I took a look at your solution. I see that you perform not many compilations, but just copy files around, run some Python UI builders, compile the protobuf stub and that's it. Is it a feature-complete Anki 2.1.44 what you obtain?

Anyway, I don't know the details about `aqt/data/web` because it's the first time I hear of Sass, I just see that there is an upstream ruleset (again, mirrored by the Anki team) that performs the generation.

Another thing: apparently, they switched to mdbook-toc for compiling the manual, so the old asciidoc recipe is no longer valid. mdbook-toc is apparently still missing from Nixpkgs, so its merge would be a minor blocker for this PR. But it's a small utility, it wouldn't be too difficult to get the PR under way.

It's crazy the level of complexity that got into this application. Plus, there aren't many Bazel builds in Nixpkgs from which we can take inspiration, so even the integration step Nix-Bazel will be an interesting one.

That being said, I think I will get back to debugging the build right now. I hope to achieve something in the coming days. If not, I'll go hunting for a Bazel enthusiast to ask for help.

**Comment by ArdaXi** [comment]
@nessdoor Well, it is a Python application at the core still, so there's not a huge amount of compilation involved, no. The Rust section does need it, see `rsbridge.nix`. You can tell it's the Sass stuff where I gave up...

I can confirm that that code is running on my system and I've been using it daily since then. UI scaling doesn't work very well (in the app settings, Wayland scaling is fine) but beyond that I've had zero issues.

**Comment by nessdoor** [comment]
If it's maintainable enough and the inputs can be overridden with not too much arcane magic, then once the manual and Sass compilation are fixed maybe you can attempt a PR? I don't think we have a reason to cope with the upstream Bazel build if not for: 1. maintainability 2. philosophy.

**Comment by ArdaXi** [comment]
Bypassing Bazel like this is bound to create maintainability issues down the line, since we're always going to be one step behind. I'm really not sure it's feasible to try and make it work within Nixpkgs though, at least in the short term.

I'd obviously like to see this get in and I'm not terribly concerned on how, but I'm afraid my available time/energy to figure out the rest of the build is a bit limited. I was just really happy when after hours of work my Anki wasn't blurry any more...

**Comment by nessdoor** [comment]
*heavy heavy curses*
https://github.com/bazelbuild/rules_rust/issues/64#issuecomment-687704649

**Comment by nessdoor** [comment]
Final update: it works.

I'll test the thing a little more, clean up the build process, try to port it to the Nixpkgs framework and then open a PR that supersedes this one.

I hope these steps will be less painful.

I am very sorry for the wait.

**Comment by samm81** [comment]
@nessdoor thanks for your work on this, I'm trying to access the IPFS hash but it doesn't seem to be on the network (atm anyways) - is there any other way to share this? maybe on github itself?

**Comment by nessdoor** [comment]
@samm81 hello, thank you for notifying me!
I put my node temporarily offline because of power and money conservation, sorry, I didn't think about the patch!
In any case, I put an updated patch at the following [gist](https://gist.github.com/nessdoor/beb96090f4a0cb246c5e067731333958). It's against the upstream release 1.47.

While I'm at it, a quick update: unfortunately, I was unable to make it run under the system Nix builder. I had to take away the handy recursive Nix invocations, but something broke in the resulting environment.
I was lazily waiting for the Nix release as to have 2.4 available in unstable, so that I could try and run the daemon with the experimental recursive invocation support enabled, but now that I think about it, I can have that already with the current unstable. Will try over the weekend.

Anyway, after applying the patch, be sure to edit the store paths inside `rules_nodejs.patch` as to reflect the hashes in _your_ Nix store. Not very reproducible, but should get you going. I've been daily-driving the resulting Anki for 2 months now (sigh, I am awful), and I can't see any issue with it.

**Comment by samm81** [comment]
@nessdoor thank you for getting back to me so quickly - I downloaded and applied the patch, but am not sure where the entry point to the build is - `find . | grep \.nix' gives:

```
./nix/node/nixpkgs_node.nix
./nix/rust/rustenv.nix
./nixpkgs-tools.nix
````

none of which look like entry points

**Comment by nessdoor** [comment]
@samm81 you can use one of the upstream entrypoints (shell scripts in the top directory and under `scripts`).
Actually, this is a Bazel build, so you need to have Bazel available.
You can see the build commands by reading the upstream entrypoints.

**Comment by nessdoor** [comment]
More than a year later, spurred by the discussion happening in NixOS/nixpkgs#78449, I decided to start investing a little bit of my free time on this, again. The latest versions of Anki have switched build system to a combination of a custom meta-builder written in Rust + Ninja.

For now, I just started writing a Nix expression that goes like this:
```
{ lib, stdenv, fetchurl, ninja, rustPlatform, ... }:

let
  src = fetchurl {
    url = "https://github.com/ankitects/anki/archive/refs/tags/${version}.tar.gz";
    sha256 = "sha256-H/DMMWZBFSTlf6qauT2hbkOXXKb2KXsx+Y7LSwvfnjI=";
  };
in stdenv.mkDerivation {
  pname = "anki";
  version = "2.1.55";
  inherit src;

  cargoDeps = rustPlatform.importCargoLock {
    lockFile = ./Cargo.lock;
    #outputHashes = {};
  };

  # Compile the custom meta-build system
  buildAndTestSubdir = "build/runner";

  nativeBuildInputs = with rustPlatform; [
    cargoSetupHook
    cargoBuildHook
    ninja
  ];

  # Set build root to $out
  BUILD_ROOT = "";
  # Activate optimizations
  RELEASE = "1";
  # Use the Python interpreter from Nixpkgs, otherwise the build system will try
  # downloading it
  PYTHON_BINARY = "";
}
```

I noticed that by setting certain environment variables, some of the build steps that automatically download stuff are inhibited. I guess our best bet is to generate a sane `build.ninja` by toying around with the custom meta-builder, and then invoke our Ninja hooks on that.

For now, consider my new efforts even less than a draft. I can't promise I will be able to commit my time to this in a stable manner.

**Comment by euank** [comment]
I picked this up over in #221229

Thank you for the initial start and thoughts, @nessdoor, I started from that for my PR, and it was helpful!

**Comment by nessdoor** [comment]
Thank you for leading this story to an end (hopefully), glad to have been of any help. Really looking forward to a fresh, sleek, source-built Anki package, finally.

**Pull request closed without merging**